qualtran/bloqs/gf_arithmetic/gf2_multiplication.py

```diff
@@ -653,3 +660,333 @@ def _binarypolynomialmultiplication() -> BinaryPolynomialMultiplication:
+        h = bb.join(h_arr[: self.n][::-1], self.qgf)
+        h = bb.add(
+            GF2MultiplyByConstantMod.from_polynomials(
+                [0, self.k], self.m_x, field_representation='int'
+            ),
+            g=h,
+        )
+        h_arr[: self.n] = bb.split(h)[::-1]
+
+        f_arr = f_arr[::-1]
+        g_arr = g_arr[::-1]
+        h_arr = h_arr[::-1]
+        f = bb.join(f_arr, self.qgf)
+        g = bb.join(g_arr, self.qgf)
+        h = bb.join(h_arr, self.qgf)
+        return {'f': f, 'g': g, 'h': h}
+
+
+@attrs.frozen
+class GF2Mul(Bloq):
+    r"""Multiplies two GF($2^n$) numbers (or binary polynomials) modulu.
+
+    Applies the transformation
+    $$
+        \ket{f}\ket{g} \rightarrow \ket{f} \ket{g} \ket{f*g \mod m(x)}
+    $$
+
+    Where the modulus $m(x)$ is the irreducible polynomial defining the galois field arithmetic.
+    The toffoli complexity is $n^{\log_2{3}}$
+
+    Args:
+        m_x: The irreducible polynomial that defines the galois field.
+        uncompute: Whether to compute or uncompute the product.
+
+    Registers:
+        f: The first number (polynomial).
+        g: The second number (polynomial).
+        h: The result.
+
+    References:
+        [Space-efficient quantum multiplication of polynomials for binary finite fields with
+            sub-quadratic Toffoli gate count](https://arxiv.org/abs/1910.02849v2) Algorithm 4.
+    """
+
+    m_x: Poly = attrs.field(converter=lambda x: x if isinstance(x, Poly) else Poly.Degrees(x))
+    uncompute: bool = False
+
+    def __attrs_post_init__(self):
+        if self.m_x.degree < 2:
+            raise ValueError(f'GF2Mul is not supported for {self.m_x}')
+
+    @cached_property
+    def n(self):
+        return int(self.m_x.degrees.max())
+
+    @cached_property
+    def gf(self):
+        return GF(2, self.n, irreducible_poly=self.m_x if self.n > 2 else None)
+
+    @cached_property
+    def qgf(self):
+        return QGF(2, self.n, irreducible_poly=self.m_x)
+
+    def adjoint(self) -> 'GF2Mul':
+        return attrs.evolve(self, uncompute=not self.uncompute)
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        # C is directional
+        side = Side.LEFT if self.uncompute else Side.RIGHT
+        return Signature(
+            [
+                Register('f', dtype=self.qgf),
+                Register('g', dtype=self.qgf),
+                Register('h', dtype=self.qgf, side=side),
+            ]
+        )
+
+    @cached_property
+    def k(self):
+        if isinstance(self.n, int):
+            return (self.n + 1) // 2
+        else:
+            return sympy.ceiling(self.n / 2)
+
+    @cached_property
+    def _gf2mulmod_impl(self) -> Bloq:
+        impl = _GF2MulImpl(self.m_x)
+        if self.uncompute:
+            return impl.adjoint()
+        return impl
+
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', f: 'Soquet', g: 'Soquet', h: Optional['Soquet'] = None
+    ) -> Dict[str, 'Soquet']:
```

Comment:
I'd personally prefer taking a `**soqs: 'SoquetT'` and simply popping the necessary registers and avoiding the need to use a `cast` and a dedicated early return based on the `if self.uncompute` condition.

Reply:
done